Tim/514 218 build clear on brand developer users of moosejs (#120)
514-labs/moose: apps/framework-docs/src/pages/building/make-changes.mdx

# Changing your data models
Change management in your data stack is critical to ensure that your data is always accurate and up to date based on your end user's expectations. Broken dashboards, reports, or machine learning models can to unhappy customers and lost revenue. MooseJS makes it easy to change your data models and ensure that your data stack is always up to date. To do so, we leverage git workflows to ensure that changes are tracked and can be rolled back if necessary.

## Adding a new data model
Adding a new data source is as simple as adding a new model to an existing or new file in the `data models` directory. MooseJS will automatically detect the new model and add it to the data stack. 

## Changing an existing data model

### During development

> Whenever you save a file, Moose will automatically detect the change and update your data stack. Under the hood, MooseJS manages the state of your stack and creates temporary tables each time you save your file. This allows you to test your changes without worrying the effects on the existing data. 

Reviewer: ```suggestion
Whenever you save a file, Moose will automatically detect the change and update your data stack locally. Under the hood, MooseJS manages the state of your stack and creates temporary tables each time you save your file. This allows you to test your changes without worrying the effects on the existing data. 
```

> Once you commit your changes, MooseJS will automatically create a new version of your data stack and turn on the migration mechanisms that ensure that historical data is present in your modified schema. The migration mechanisms assume that the initial state of your data starts at the initial state of your branch. This limits the number of active migrations that are running in your stack to optimize performance and costs.

Reviewer: Not sure what you mean by that. Why isn't the migration base state the previous commit?

The advantage of using a commit is that it is the same simple predictable process on the `main` branch or on feature branches

Author: Because you'd need a new infra + migration mechanism for each step no?

Reviewer: Yeah - as long as we have a way to prevent it from getting too large on disk that should be ok no?

> If you already have a version of the SDK instrumented, that version will remain live and be marked for deprecation. This allows you to test your changes in without needing to update all your exiting instrumentation.

Reviewer: ```suggestion
If you are already using a version of the sdk, that version will remain live and be marked for deprecation. This allows you to test your changes in without needing to update all your exiting instrumentation.
```

### In production
Once you're ready to deploy your changes to production, you can merge your branch into the `main` branch. MooseJS will automatically detect the merge and create a new version of your data stack with all the migrations taken care of behind the scenes while ensuring that your SDKs historical SDKs continue to work.

## Deleting a data model

> Deleting a data model is as simple as deleting the model from the file. MooseJS will automatically flag the model for deprecation and create a new version of your data stack. This allows you to remove data models that are no longer needed without worrying about breaking systems that depend on your data service. Once no more data is being pushed to your service or consumed, Moose will automatically clean up your infrastructure.

Reviewer: I am not sure this should be the behavior.

If I understand correctly, In this case we will end up with the code in the repo that doesn't reflect the production environment, where, for some time, the infra will have data flow through with no code or model in the code.

We could recommend here the strangler pattern in which people would remove the model once nothing is consuming the data and then delete the model in the code. When the model is deleted, then the infrastructure would be deprovioned immediately

Author: Ok, let's talk more about this and fold it into the PRD. I'm a bit worried about hard deletes here and the potential implications.

Reviewer: That's a fair worry. If we go with the git based approach though - people will need to do reviews in order to merge a deletion - as such that will be reviewed before it gets executed in prod

Does that mitigate the worry?
